SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestCrossServerRPC/SpatialTestCrossServerRPC.cpp

```diff
@@ -33,43 +46,118 @@ void ASpatialTestCrossServerRPC::PrepareTest()
-	TArray<FVector> CubesLocations;
-	CubesLocations.Add(FVector(250.0f, 250.0f, 75.0f));
-	CubesLocations.Add(FVector(250.0f, -250.0f, 75.0f));
-	CubesLocations.Add(FVector(-250.0f, -250.0f, 75.0f));
-	CubesLocations.Add(FVector(-250.0f, 250.0f, 75.0f));
-
-	AddStep(TEXT("EnsureSpatialOS"), FWorkerDefinition::Server(1), nullptr, [this]() {
+	// Pre-test checks
+	AddStep(TEXT("Pre-test check"), FWorkerDefinition::Server(1), nullptr, [this]() {
 		USpatialNetDriver* SpatialNetDriver = Cast<USpatialNetDriver>(GetNetDriver());
+		SpatialNetDriver = Cast<USpatialNetDriver>(GetNetDriver());
 
 		if (SpatialNetDriver == nullptr || SpatialNetDriver->LoadBalanceStrategy == nullptr)
 		{
 			FinishTest(EFunctionalTestResult::Error, TEXT("Test requires SpatialOS enabled with Load-Balancing Strategy"));
 		}
 		else
 		{
-			FinishStep();
+			CheckInvalidEntityID(LevelCube);
 		}
 	});
 
-	for (int i = 1; i <= 4; ++i)
+	// Startup actor tests
+	//
+	// Repro the error case of the entity ID being incorrectly allocated on a non-auth server - expect warnings
+	if (HasAuthority())
+	{
+		// Expect this warning from each non-authoritative server
+		AddExpectedLogError(TEXT("that is replicated and exists on another worker, but we haven't received from runtime. Dropping RPC."), 3,
+							false);
+	}
+	AddStep(TEXT("Startup actor tests: repro error case for entity ID after RPC on non-auth server"), FWorkerDefinition::AllServers,
+			nullptr, [this]() {
+				int LocalWorkerId = GetLocalWorkerId();
+
+				// These specific steps were needed to recreate an error of the entity ID being incorrectly allocated on a non-auth server.
+				// The level actor is located on server 4 in the map and is not replicated initially. Therefore, all actors are initially
+				// authoritative. First we turn on replication on the level actor, then we change it to non-auth and finally send a cross
+				// server RPC. It is the ProcessRPC call that was causing an incorrect entity ID to be allocated in this specific case but
+				// this has now been fixed and we are expecting a warning error in this case.
+				if (LocalWorkerId < 4)
+				{
+					LevelCube->TurnOnReplication();
+					LevelCube->SetNonAuth();
+					LevelCube->CrossServerTestRPC(LocalWorkerId);
+				}
+				FinishStep();
+			});
+
+	AddStep(TEXT("Startup actor tests: Post-RPC entity ID check"), FWorkerDefinition::AllServers, nullptr, nullptr,
+			[this](float DeltaTime) {
+				CheckInvalidEntityID(LevelCube);
+			});
+
+	// Normal case
+	AddStep(TEXT("Startup actor tests: Auth server - Set replicated"), FWorkerDefinition::Server(4), nullptr, [this]() {
+		LevelCube->TurnOnReplication();
+		FinishStep();
+	});
+
+	AddStep(TEXT("Startup actor tests: Auth server - Record entity id"), FWorkerDefinition::Server(4), nullptr, [this]() {
+		LevelCube->RecordEntityId();
+		FinishStep();
+	});
+
+	AddStep(TEXT("Startup actor tests: Post-Auth entity ID check"), FWorkerDefinition::AllServers, nullptr, nullptr,
+			[this](float DeltaTime) {
```

Comment:
Likewise here

Reply:
Good point, waiting for IsActorReady before executing this step.